### PR TITLE
[litmus] Add support for -variant self in -mode presi

### PR DESCRIPTION
**Pull request by relokin**

### CODE REVIEW

**Comment by relokin** [comment]
Converted to draft to allow me some more time to test this, but I think all the functionality is there.

Also please note that the 1st patch in this PR fixes a compiler error in -mode kvm without -variant self.

File: litmus/preSi.ml [hunk 1]
```diff
@@ -1355,12 +1355,14 @@ module Make
           O.o "}" ;
           O.o "" ;
           O.o "static void vars_free(vars_t *_vars) {" ;
-          let open OutUtils in
-          List.iter
-            (fun (n,_) ->
-              O.fi "free_pages(_vars->%s);" (fmt_code n))
-            test.T.code ;
-          O.o "}" ;        
+          if do_self then begin
```
Comment (maranget) [inline litmus/preSi.ml:1367]: Oups, thanks

Comment (relokin) [reply litmus/preSi.ml:1367]: How can I help get this in? Should I create a separate PR? Or do you want to cherry-pick it by hand?

Comment (maranget) [reply litmus/preSi.ml:1367]: A separate PR would be the best method I guess. If you do it, I'll merge it immediately.

**Comment by relokin** [comment]
Added one more commit that fixes issue #593

**Comment by relokin** [comment]
hey @maranget, I think this is ready. Please let me know if you have any questions, or I can do something to help with the review.

**Review by maranget — COMMENTED** [review]
Hi @relokin, nice work. I have tried quite a few variation of modes and options and it works :).

It is some king of achievement because the current code results from combining various features on labels and instructions. As a consequence , the code is somehow intricated and would benefit from a simplification. Let us spare this for a future PR....

File: litmus/libdir/_aarch64/presi-self.c [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+}
+
+static size_t code_size(ins_t *p,int skip) {
+  return (find_ins(getret(), p, skip) + 1) * sizeof(ins_t);
+}
+
+static void code_init(void *code, void *src, size_t sz)
+{
+  memcpy(code, src, sz);
+  litmus_icache_sync((unsigned long)code, (unsigned long)code + sz);
```
[the remark this answers is not in the document]
Comment (relokin) [reply litmus/libdir/_aarch64/presi-self.c:35]: thanks!

File: litmus/preSi.ml [hunk 3]
```diff
@@ -1461,6 +1489,11 @@ module Make
         if do_self then
           O.fii "code_init(%s, %s, _vars->%s);" (LangUtils.code_fun_cpy proc)
             (OutUtils.fmt_code proc) (OutUtils.fmt_code_size proc) ;
+        if need_labels test then begin
+            if proc = 0 then
+            O.oii "init_labels(_vars);" ;
```
[the remark this answers is not in the document]
Comment (relokin) [reply litmus/preSi.ml:1494]: Actually, I am having 2nd thoughts, let see if I can do some clean-up as part of this PR. Thanks for review this!

[the remark this answers is not in the document]
Comment (maranget) [reply litmus/preSi.ml:1494]: For your information. I am leaving on vacations for two weeks on saturday...

File: litmus/libdir/_aarch64/presi-self.c [hunk 4]
```diff
@@ -0,0 +1,36 @@
...
+/* abiding by the rules of distribution of free software. You can use,      */
+/* modify and/ or redistribute the software under the terms of the CeCILL-B */
+/* license as circulated by CEA, CNRS and INRIA at the following URL        */
+/* "http://www.cecill.info". We also give a copy in LICENSE.txt.            */
+/****************************************************************************/
+/* Authors:                                                                 */
+/* Nikos Nikoleris, Arm Limited.                                            */
+/****************************************************************************/
+
+static void litmus_icache_sync(unsigned long vaddr, unsigned long vaddr_end)
```
[the remark this answers is not in the document]
Comment (maranget) [reply litmus/libdir/_aarch64/presi-self.c:20]: Argl untested code...

**Comment by relokin** [comment]
I've pushed the uintptr_t change and it seems that we're all in agreement to merge this as-is (@jalglave, @maranget)?

In parallel, I'm working on changing the labels to offset and a couple of other clean-ups I would like to do which will go into a separate PR for after @maranget you're back.

**Comment by jalglave** [comment]
> I've pushed the uintptr_t change and it seems that we're all in agreement to merge this as-is (@jalglave, @maranget)?
> 

Yes I'm happy with that as discussed with you both. If @maranget is happy too we can go ahead.
Thanks you both very much.
Jade

**Comment by maranget** [comment]
All squashed and rebased for keeping a neat history, and ... merged! Thanks.

**Pull request merged**